### ISSUE TITLE
Move from Travis CI to new CI provider

### ISSUE THREAD

**Issue by brianlheim** [supercollider/supercollider#5252]
## Motivation

https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing

https://docs.travis-ci.com/user/migrate/open-source-repository-migration/#frequently-asked-questions

as i understand it, this means:
- travis-ci.org, what we currently use, will shut down at the end of the year
- travis-ci's "pro" service, travis-ci.com, is getting a new pricing model with the following anti-features:
  - we need credits to run mac builds (50 credits / 1 minute of build time)
  - travis will only give us 10k credits, non-renewing
  - we need to ask for more, and they may not give it to us

this is a serious problem and we should plan to move off Travis CI within the next month or two. thanks to @claremacrae for making us aware!

## Description of Proposed Feature

we need to find a new CI provider for linux and mac builds. some soft requirements are:
- support our current beefed-up build matrix if possible
- completes all builds in under 40 min (much easier the more control we have over the environment)
- integrates with github
- self-hosted is an option, as long as uptime is guaranteed and multiple people can administer build machines

maybe, potentially consider options that cost up to $50 / month. worth considering if the options are very limited and a paid solution would be the most stable.

## Plan for Implementation

@joshpar is currently looking into this.

i think it's a long shot, but we could email travis to ask for the credits we'd need, earlier in slack i estimated this to be on the order of 250k per month based on our current usage, but someone else should check my math.

some CI providers i know of that may be worth looking into:
- appveyor (cloud-hosted for mac i think?)
- jenkins
- circle CI
- github actions

**Comment by brianlheim** [comment]
i did a quick survey of some options and i think it may just be time to bite the bullet and switch to github actions. there are virtually no limits on build times, finer-grained control over build machines, and we would be able to move both our appveyor and travis jobs to it. since we only have around 6 weeks to make this change, i'd say we should make a final decision by the end of the month.

**Comment by brianlheim** [comment]
mac stadium offers a free hosted mac mini for OSS projects: https://www.macstadium.com/opensource

only one concurrent job for mac isn't the worst thing in the world. currently, our mac builds altogether take about 45 minutes with 2 concurrent builds as provided by travis. but, almost all of that time is spent updating homebrew and installing packages. if we were able to cache those things, i would expect to see total job time drop from 45 minutes to about the same as linux builds, 12-16 minutes for 4 jobs.

**Comment by brianlheim** [comment]
some numbers i used for estimating the time we would need:
- 4 mac jobs per build
- 14 linux jobs per build
- ~10-15 min per mac job (but could reduce this significantly, see above)
- 2-4 min per linux job
- 100-250 builds per month

**Comment by claremacrae** [comment]
FWIW I really like GitHub Actions.

* They run much faster than Travis and AppVeyor. So you may find that you don't need as much time as estimates based on Travis would suggest.
* They run more steps in parallel, so elapsed time is shorter
* The integration in to GitHub is really nice (clicking on Actions in the "Toolbar" takes you straight to the builds - no fiddling around trying to find a checkmark or cross to click on)
* When you download the full build log, you get a link to the exact machine setup details, which is super useful tracking down failures caused by change of state. You can effectively diff the state of machines between a failing and passing build
* You can set up multiple .yml files, for different types of builds (like one to check formatting, one to do C++ builds, one to run python tests....) This gives separation of concerns both when maintaining the YAML, and when viewing build logs
* The way you can reuse other people's actions is really nice - it keeps the YAML shorter.

**Comment by claremacrae** [comment]
Wow, taking the upper bounds in Brian's numbers, I calculate the most GitHub minutes you would use in a month would be 155,000 - and the free limit is 2,000!

**Comment by claremacrae** [comment]
See [SuperCollider build estimates.xlsx](https://github.com/supercollider/supercollider/files/5559195/SuperCollider.build.estimates.xlsx) for my working.

**Comment by brianlheim** [comment]
> Wow, taking the upper bounds in Brian's numbers, I calculate the most GitHub minutes you would use in a month would be 164,000 - and the free limit is 2,000!

yeah, and this seems to be a common theme across CI providers. i think one option we should seriously consider is using a service like GitHub Actions or Circle CI with a self-hosted macOS runner from mac stadium. the macOS credit allotments seem to be the most restrictive part of each plan, so if we don't have to worry about those limits we will probably have more options.

**Comment by claremacrae** [comment]
I wonder if anything radical can be done to reduce the build time requirements?

For example....

Have you reviewed all the current builds to see if any are not necessary?

For the ones that remain, can any be combined in to one job, so the set-up time is shared across several different build configs?

**Comment by claremacrae** [comment]
I wonder if it's worth doing an initial experiment on GitHub, of 1 Mac  build and 1 Linux build, to get data on the elapsed times.

**Comment by claremacrae** [comment]
I wonder if anything radical can be done to reduce the build times - can anything be pruned from the build steps?

**Comment by brianlheim** [comment]
> I wonder if anything radical can be done to reduce the build times - can anything be pruned from the build steps or the code?

yes, a few things. if we have access to an ubuntu 20 environment then we can probably move the syslibs job from macOS to linux. if we have more control over the build machines then we can eliminate the costly homebrew update and install phases. But this will still give us 2-3 minutes per job and 3 jobs per build in the absolute best conditions. Even if we pare it down to one macOS build, we would need about 300-500 minutes of mac build time per month. this is still nowhere near github action's credit allotment -- since we need to spread the 2000 credits across both mac and linux -- and just barely below Circle CI's, which would leave us with no room for growth or mistakes.

**Comment by brianlheim** [comment]
> I wonder if it's worth doing an initial experiment on GitHub, of 1 Mac build and 1 Linux build, to get data on the elapsed times.

i'd be interested to see data on this too for each service we're considering.

**Comment by claremacrae** [comment]
> i'd be interested to see data on this too for each service we're considering.

Yes, that's a much better way of doing it...

**Comment by brianlheim** [comment]
@dyfer and i just discussed this, here are some notes from that conversation:
- we could temp use appveyor for macOS since we already have an account there and it's free, and since we can apparently continue to run linux jobs on travis's .com service after travis's .org is shut down
- we could have the 'full set' of jobs run on a cron job and not every PR. needs analysis as to long-term sustainability as this doesn't give much room for growth or error

**Comment by claremacrae** [comment]
From https://docs.github.com/en/free-pro-team@latest/github/setting-up-and-managing-billing-and-payments-on-github/about-billing-for-github-actions

> **GitHub Actions usage is free for public repositories and self-hosted runners.** For private repositories, each GitHub account receives a certain amount of free minutes and storage, depending on the product used with the account. By default, your account will have a spending limit of $0, which prevents additional usage of minutes or storage after you reach these limits. If you increase your spending limit above the default of $0, you will be billed for any minutes or storage beyond the limits, also called overages. GitHub charges usage to the account that owns the repository where the workflow is run. Any coupons on your account do not apply to GitHub Actions overages.

All the calculations that Brian and I did were relevant for private repos.

How many private repos does SuperColider need builds for?

If, as I suspect, the answer is 0, SuperColider will be fine with GHA.

(This was confirmed by Llewellyn Falco, whose public repos use way more than the limits we were calculating on....)

**Comment by brianlheim** [comment]
!!! oh wow! i didn't know this at all, thanks @claremacrae !

**Comment by brianlheim** [comment]
"  free for public repositories and self-hosted runners" -- does this mean free for all public repos AND people using self-hosted runners, or people who have public repos that use self-hosted runners only? (hope that question is clear)

**Comment by claremacrae** [comment]
So I stand by my list of reasons above why GHA is awesome... and would like to very gently steer you in that direction!

**Comment by brianlheim** [comment]
:))) i just want to confirm, we can use GHA without self-hosting and it will be unlimited + free, is that correct? if so, it's the clear best choice as far as i'm concerned

**Comment by claremacrae** [comment]
If your repos are public, GHA is free.

If you are self-hosting your runners, GHA is free.

If your repos are private, and run on GH infrastructure, then the limits and fees kick in.

**Comment by claremacrae** [comment]
> :))) i just want to confirm, we can use GHA without self-hosting and it will be unlimited + free, is that correct? if so, it's the clear best choice as far as I'm concerned

Yep - spot on.

**Comment by brianlheim** [comment]
ahhh thank you so much! this is great.

**Comment by brianlheim** [comment]
also just to confirm:

> How many private repos does SuperColider need builds for?

we do not use any private repos, and i don't think we would ever use them out of principle, unless it was absolutely necessary for some unforeseen privacy or security reason.

**Comment by joshpar** [comment]
rad

**Comment by brianlheim** [comment]
so IIUC we are moving to GHA and don't need to continue searching for other options. i would note that to me this very much feels like the second step of "embrace, extend, extinguish", but as an unfunded open source project our choices are regrettably limited here.

if we can also move from appveyor to GHA without much hassle, i think we should consider it if it improves our overall CI build time.

@dyfer and @claremacrae have both said they are interested to help with the transition, thanks to you both! i am also happy to help if needed.

if there are no concerns with this plan, i would like to close this and open a new ticket to discuss the move to GHA specifically.

**Comment by joshpar** [comment]
Agreed

/*
Josh Parmenter
www.realizedsound.net/josh
*/
